### PR TITLE
Feature/emotion to event

### PR DESCRIPTION
**Pull request by Nir-P**
### This feature adds emotion to an event based on its title and content

The main function "get_emotion" returns the emotion HTML code based on the event title and content.
The title and content were given weights as we set in the global variables.
Additionally, we set the significance level that above it the emotion will appear.

**Changes in models:**
The event table now has a new column: emotion
this column returns a string (HTML code) and may be null.

**The following emotion are:**

😃 Happy - &#128515
🙁 Sad - &#128577
😠 Angry - &#128544
😱 Fear - &#128561
😮 Surprise - &#128558

### CODE REVIEW

File: app/internal/emotion.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def dominant_emotion(title: str, content: str) -> Union[List]:
+    """
+    get text from event title and content and return
+    the dominant emotion, emotion score and smiley code
+    """
+    dominant_emotion = [None, 0, None]
+    duplicate_dominant_flag = 0
+    title_emotion = te.get_emotion(title)
```
Comment (yammesicka) [inline app/internal/emotion.py:24]: Prefer to choose on, as te.get_emotion might be a heavy computing calculation

File: app/internal/emotion.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    if content is not None:
+        content_emotion = te.get_emotion(content)
+    for emotion, code in EMOTIONS:
+        if content not in (None, "", " "):
```
Comment (yammesicka) [inline app/internal/emotion.py:28]: `if not content or not content.strip()` might find many more cases

File: app/internal/emotion.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    get text from event title and content and return
+    the dominant emotion, emotion score and smiley code
+    """
+    dominant_emotion = [None, 0, None]
+    duplicate_dominant_flag = 0
+    title_emotion = te.get_emotion(title)
+    if content is not None:
+        content_emotion = te.get_emotion(content)
+    for emotion, code in EMOTIONS:
```
Comment (yammesicka) [inline app/internal/emotion.py:27]: Maybe we should export the content of this for to another function? :)

File: app/internal/emotion.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            ("Fear", "&#128561"),
+            ("Surprise", "&#128558")]
+
+
+def dominant_emotion(title: str, content: str) -> Union[List]:
+    """
+    get text from event title and content and return
+    the dominant emotion, emotion score and smiley code
+    """
+    dominant_emotion = [None, 0, None]
```
Comment (yammesicka) [inline app/internal/emotion.py:22]: Consider using  NamedTuple instead, to make the code more obvious :)

File: app/internal/emotion.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+EMOTIONS = [("Happy", "&#128515"),
+            ("Sad", "&#128577"),
+            ("Angry", "&#128544"),
+            ("Fear", "&#128561"),
+            ("Surprise", "&#128558")]
+
+
+def dominant_emotion(title: str, content: str) -> Union[List]:
```
Comment (yammesicka) [inline app/internal/emotion.py:17]: List of...? :)

File: app/internal/emotion.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import text2emotion as te
+from typing import List, Union
+
+from app.config import (
+    CONTENT_WEIGHTS,
+    LEVEL_OF_SIGNIFICANCE,
+    TITLE_WEIGHTS)
+
+
+EMOTIONS = [("Happy", "&#128515"),
```
Comment (yammesicka) [inline app/internal/emotion.py:10]: Might as well be a dictionary?

File: app/internal/emotion.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+def is_emotion_above_significance(dominant_emotion: Union[List],
+                                  significance=LEVEL_OF_SIGNIFICANCE) -> bool:
+    """
+    get the dominant emotion, emotion score and smiley code
+    and check if the emotion score above the constrain we set
+    """
+    return dominant_emotion[1] >= significance
+
+
+def get_html_smiley(dominant_emotion: Union[List]) -> Union[str, None]:
```
Comment (yammesicka) [inline app/internal/emotion.py:52]: Why is there a separate function to one of the emoticons?

File: app/internal/emotion.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    return dominant_emotion
+
+
+def is_emotion_above_significance(dominant_emotion: Union[List],
+                                  significance=LEVEL_OF_SIGNIFICANCE) -> bool:
+    """
+    get the dominant emotion, emotion score and smiley code
+    and check if the emotion score above the constrain we set
+    """
+    return dominant_emotion[1] >= significance
```
Comment (yammesicka) [inline app/internal/emotion.py:49]: Why specifically 1?

**Review by yammesicka — COMMENTED** [review]
Great start! Please clarify the code a bit to make it easier to skim through :)

File: app/internal/emotion.py (hunk 1 above)
Comment (Nir-P) [reply app/internal/emotion.py:29]: could you explain what do you mean to choose on?
thanks!

File: app/internal/emotion.py (hunk 7 above)
Comment (Nir-P) [reply app/internal/emotion.py]: changed the function name to get_html_emoticon (I mixed up the word smiley with emoticon)

File: app/internal/emotion.py (hunk 8 above)
Comment (Nir-P) [reply app/internal/emotion.py]: I used NamedTuple, I'm hoping it is clear now

File: app/internal/emotion.py (hunk 2 above)
Comment (Nir-P) [reply app/internal/emotion.py]: Added content.strip() thanks!

File: app/internal/emotion.py (hunk 3 above)
Comment (Nir-P) [reply app/internal/emotion.py]: did it :)
split to `get_weights `and `score_comp`

File: app/internal/emotion.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    TITLE_WEIGHTS)
+
+
+EMOTIONS = {"Happy": "&#128515",
+            "Sad": "&#128577",
+            "Angry": "&#128544",
+            "Fear": "&#128561",
+            "Surprise": "&#128558"}
+
+Emoticon = namedtuple("Emoticon",
```
Comment (yammesicka) [inline app/internal/emotion.py:18]: Prefer using typing.NamedTuple

File: app/internal/emotion.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    score comparison between emotions.
+    returns the dominant and if equals we flag it
+    """
+    if emotion_score > DominantEmotion.emotion_score:
+        flag = 0
+        DominantEmotion = Emoticon(emotion, emotion_score, code)
+    elif emotion_score == DominantEmotion.emotion_score:
+        flag = 1
+    return (DominantEmotion, flag)
```
Comment (yammesicka) [inline app/internal/emotion.py:41]: What if emotion_score < DominantEmotion.emotion_score?

File: app/internal/emotion.py (hunk 1 above)
Comment (yammesicka) [reply app/internal/emotion.py:52]: Probably meant "prefer to put on the `else` clause", but now I see you need them both ^^"

File: app/internal/emotion.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+
+EMOTIONS = {"Happy": "&#128515",
+            "Sad": "&#128577",
+            "Angry": "&#128544",
+            "Fear": "&#128561",
+            "Surprise": "&#128558"}
+
+Emoticon = namedtuple("Emoticon",
+                      ["dominant_emotion", "emotion_score", "emoticon_code"])
```
Comment (yammesicka) [inline app/internal/emotion.py:19]: Prefer: `dominant`, `score`, `code`

File: app/internal/emotion.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        if no_content:
+            emotion_score = get_weights(emotion, title_emotion)
+        else:
+            emotion_score = get_weights(emotion, title_emotion,
+                                        content_emotion)
+        score_comparison = score_comp(emotion_score, DominantEmotion, emotion,
+                                      code, duplicate_dominant_flag)
+        DominantEmotion, duplicate_dominant_flag = [*score_comparison]
+    if duplicate_dominant_flag:
+        return Emoticon(None, 0, None)
```
Comment (yammesicka) [inline app/internal/emotion.py:66]: Prefer adding the field names for clarity: `Emoticon(dominant=None, ...)`

File: app/internal/emotion.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    return DominantEmotion.emoticon_code
+
+
+def get_emotion(title: str, content: str) -> Union[str, None]:
+    """
+    The main function checks what the dominant emotion
+    and if thr dominant emotion above the constrain we set
+    return the emoticon code
+    """
+    Dominant = dominant_emotion(title, content)
```
Comment (yammesicka) [inline app/internal/emotion.py:90]: Why `Dominant` and not `dominant`?

File: app/internal/emotion.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+                                  LEVEL_OF_SIGNIFICANCE) -> bool:
+    """
+    get the dominant emotion, emotion score and emoticon code
+    and check if the emotion score above the constrain we set
+    """
+    return DominantEmotion.emotion_score >= significance
+
+
+def get_html_emoticon(DominantEmotion: Emoticon) -> Union[str, None]:
+    return DominantEmotion.emoticon_code
```
Comment (yammesicka) [inline app/internal/emotion.py:81]: Why DominantEmotion and not dominant_emotion?

File: app/internal/emotion.py (hunk 13 above)
Comment (yammesicka) [reply app/internal/emotion.py:90]: Maybe `get_dominant_emotion`? (functions should be a verb)

**Review by yammesicka — COMMENTED** [review]
Cool! Thanks for addressing the issues from the last time. I've added few other insights :)

File: app/internal/emotion.py (hunk 10 above)
Comment (Nir-P) [reply app/internal/emotion.py:41]: the DominantEmotion will not change

File: app/internal/emotion.py (hunk 14 above)
Comment (Nir-P) [reply app/internal/emotion.py]: fixed it

File: app/internal/emotion.py (hunk 9 above)
Comment (Nir-P) [reply app/internal/emotion.py]: Thanks

File: app/internal/emotion.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+
+def score_comp(emotion_score: float, dominant_emotion: Emoticon, emotion: str,
+               code: str, flag: int) -> Tuple[Emoticon, int]:
+    """
+    score comparison between emotions.
+    returns the dominant and if equals we flag it
+    """
+    if emotion_score > dominant_emotion.score:
+        flag = 0
```
Comment (yammesicka) [inline app/internal/emotion.py:35]: Why not True/False?

File: app/internal/emotion.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    return (title_emotion[emotion] * TITLE_WEIGHTS +
+            content_emotion[emotion] * CONTENT_WEIGHTS)
+
+
+def score_comp(emotion_score: float, dominant_emotion: Emoticon, emotion: str,
+               code: str, flag: int) -> Tuple[Emoticon, int]:
+    """
+    score comparison between emotions.
+    returns the dominant and if equals we flag it
+    """
```
Comment (yammesicka) [inline app/internal/emotion.py:33]: What if `emotion_score < dominant_emotion.score`? 

`flag` is not defined

File: app/internal/emotion.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    if emotion_score > dominant_emotion.score:
+        flag = 0
+        dominant_emotion = Emoticon(dominant=emotion, score=emotion_score,
+                                    code=code)
+    elif emotion_score == dominant_emotion.score:
+        flag = 1
+    return (dominant_emotion, flag)
```
Comment (yammesicka) [inline app/internal/emotion.py:40]: Prefer NamedTuple

File: app/internal/emotion.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+            "Sad": "&#128577",
+            "Angry": "&#128544",
+            "Fear": "&#128561",
+            "Surprise": "&#128558"}
+
+Emoticon = NamedTuple("Emoticon", [("dominant", str), ("score", float),
+                      ("code", str)])
+
+
+def get_weights(emotion: str, title_emotion: Dict[str, float],
```
Comment (yammesicka) [inline app/internal/emotion.py:20]: Maybe `get_weight`?

File: app/internal/emotion.py [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    return (dominant_emotion, flag)
+
+
+def get_dominant_emotion(title: str, content: str) -> Emoticon:
+    """
+    get text from event title and content and return
+    the dominant emotion, emotion score and emoticon code
+    """
+    dominant_emotion = Emoticon(dominant=None, score=0, code=None)
+    no_content = 1
```
Comment (yammesicka) [inline app/internal/emotion.py:49]: Prefer:

`has_content = False`

File: app/internal/emotion.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    duplicate_dominant_flag = 0
+    title_emotion = te.get_emotion(title)
+    if content is not None and content.strip() != "":
+        content_emotion = te.get_emotion(content)
+        no_content = 0
+    for emotion, code in EMOTIONS.items():
+        if no_content:
```
Comment (yammesicka) [inline app/internal/emotion.py:56]: Prefer the DRY approach:

```
weight_parameters = [emotion, title_emotion]
if has_content:
    weight_parameters.append(content_emotion)
emotion_score = get_weights(*weight_parameters)
```

**Review by yammesicka — CHANGES_REQUESTED** [review]
Great progress! Added few additional insights.

File: app/internal/emotion.py (hunk 18 above)
Comment (Nir-P) [reply app/internal/emotion.py]: Done

File: app/internal/emotion.py (hunk 19 above)
Comment (Nir-P) [reply app/internal/emotion.py]: fixed it

File: app/internal/emotion.py (hunk 20 above)
Comment (Nir-P) [reply app/internal/emotion.py]: I liked it, thanks!

File: app/internal/emotion.py (hunk 17 above)
Comment (Nir-P) [reply app/internal/emotion.py]: done

File: app/internal/emotion.py (hunk 15 above)
Comment (Nir-P) [reply app/internal/emotion.py]: changed it

File: app/internal/emotion.py (hunk 16 above)
Comment (Nir-P) [reply app/internal/emotion.py:34]: when `emotion_score < dominant_emotion.score`  the input flag will be returned

**Review by yammesicka — APPROVED** [review]

**Pull request merged**